File: sdk/python/creation_os.py
"""Thin HTTP client for cos-serve (σ-gate API).

Requires: pip install requests
Server: build with `make cos-serve`, run `./cos-serve --port 3001` or `cos serve`.
"""

from __future__ import annotations

import os
from typing import Any, Dict, Optional

try:
    import requests
except ImportError as e:  # pragma: no cover
    raise ImportError(
        "creation_os requires the 'requests' package: pip install requests"
    ) from e


class CreationOS:
    """JSON client for POST /v1/gate, POST /v1/verify, GET /v1/health."""

    def __init__(
        self,
        host: str = "localhost",
        port: Optional[int] = None,
        *,
        base_url: Optional[str] = None,
        timeout_s: float = 600.0,
    ) -> None:
        if base_url is not None:
            self.base = base_url.rstrip("/")
        else:
            if port is None:
                port = int(os.environ.get("COS_SERVE_PORT", "3001"))
            self.base = f"http://{host}:{port}"
        self.timeout_s = timeout_s

    def gate(self, prompt: str, model: str = "gemma3:4b") -> Dict[str, Any]:
        r = requests.post(
            f"{self.base}/v1/gate",
            json={"prompt": prompt, "model": model},
            timeout=self.timeout_s,
        )
        r.raise_for_status()
        return r.json()

    def verify(self, text: str, model: str = "gemma3:4b") -> Dict[str, Any]:
        r = requests.post(
            f"{self.base}/v1/verify",
            json={"text": text, "model": model},
            timeout=self.timeout_s,
        )
        r.raise_for_status()
        return r.json()

    def health(self) -> Dict[str, Any]:
        r = requests.get(f"{self.base}/v1/health", timeout=30.0)
        r.raise_for_status()
        return r.json()

    def audit(self, audit_id: str) -> Dict[str, Any]:
        r = requests.get(f"{self.base}/v1/audit/{audit_id}", timeout=30.0)
        r.raise_for_status()
        return r.json()

File: sdk/js/creation-os.js
/**
 * Thin HTTP client for cos-serve (Node 18+ for global fetch).
 *
 * Usage:
 *   const { CreationOS } = require('./creation-os.js');
 *   const cos = new CreationOS();
 *   const r = await cos.gate('What is 2+2?');
 */

function defaultPort() {
  const p = process.env.COS_SERVE_PORT;
  return p ? parseInt(p, 10) : 3001;
}

class CreationOS {
  /**
   * @param {{ host?: string, port?: number, baseUrl?: string, timeoutMs?: number }} [opts]
   */
  constructor(opts = {}) {
    if (opts.baseUrl) {
      this.base = opts.baseUrl.replace(/\/$/, '');
    } else {
      const host = opts.host || 'localhost';
      const port = opts.port != null ? opts.port : defaultPort();
      this.base = `http://${host}:${port}`;
    }
    this.timeoutMs = opts.timeoutMs != null ? opts.timeoutMs : 600000;
  }

  async _json(method, path, bodyObj) {
    const ctrl = new AbortController();
    const id = setTimeout(() => ctrl.abort(), this.timeoutMs);
    try {
      const init = {
        method,
        headers: { 'Content-Type': 'application/json' },
        signal: ctrl.signal,
      };
      if (bodyObj != null) init.body = JSON.stringify(bodyObj);
      const res = await fetch(`${this.base}${path}`, init);
      const text = await res.text();
      let data;
      try {
        data = text ? JSON.parse(text) : {};
      } catch {
        throw new Error(`cos-serve: non-JSON response (${res.status}): ${text.slice(0, 200)}`);
      }
      if (!res.ok) {
        const err = new Error(data.error || `HTTP ${res.status}`);
        err.status = res.status;
        err.body = data;
        throw err;
      }
      return data;
    } finally {
      clearTimeout(id);
    }
  }

  async gate(prompt, model = 'gemma3:4b') {
    return this._json('POST', '/v1/gate', { prompt, model });
  }

  async verify(text, model = 'gemma3:4b') {
    return this._json('POST', '/v1/verify', { text, model });
  }

  async health() {
    return this._json('GET', '/v1/health', null);
  }

  async audit(auditId) {
    return this._json('GET', `/v1/audit/${encodeURIComponent(auditId)}`, null);
  }
}

module.exports = { CreationOS };
